Core/Core/Views/PagesViewController.swift

```diff
@@ -0,0 +1,195 @@
+    }
+
+    public override func viewDidLoad() {
+        scrollView.alwaysBounceHorizontal = true
+        scrollView.delegate = self
+        scrollView.isPagingEnabled = true
+        scrollView.showsHorizontalScrollIndicator = false
+        scrollView.showsVerticalScrollIndicator = false
+
+        embedPage(currentPage, at: 0)
+    }
+
+    public override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+        layout()
+    }
+
+    private func getPage(onLeft: Bool) -> UIViewController? {
+        let isBefore = view.effectiveUserInterfaceLayoutDirection == .leftToRight ? onLeft : !onLeft
+        return isBefore
+            ? dataSource?.pagesViewController(self, pageBefore: currentPage)
+            : dataSource?.pagesViewController(self, pageAfter: currentPage)
+    }
+
+    public func scrollViewWillBeginDragging(_ scrollView: UIScrollView) {
+        if let left = getPage(onLeft: true) {
+            leftPage?.unembed()
+            embedPage(left, at: 0)
+            if leftPage == nil {
+                scrollView.contentOffset.x += scrollView.frame.width
+            }
+            leftPage = left
+        }
+        if let right = getPage(onLeft: false) {
+            rightPage?.unembed()
+            embedPage(right, at: leftPage == nil ? 1 : 2)
+            rightPage = right
+        }
+        layout()
+    }
+
+    public func scrollViewDidScroll(_ scrollView: UIScrollView) {
+        let visible = CGRect(
+            x: scrollView.contentOffset.x,
+            y: 0,
+            width: scrollView.frame.width,
+            height: scrollView.frame.height
+        )
+        var newValue: [UIViewController] = []
+        if let left = leftPage, left.view.frame.intersects(visible) {
+            newValue.append(left)
+        }
+        if currentPage.view.frame.intersects(visible) {
+            newValue.append(currentPage)
+        }
+        if let right = rightPage, right.view.frame.intersects(visible) {
+            newValue.append(right)
+        }
+        guard newValue != showing else { return }
+        showing = newValue
+        delegate?.pagesViewController?(self, isShowing: showing)
+    }
+
+    public func scrollViewWillEndDragging(_ scrollView: UIScrollView, withVelocity velocity: CGPoint, targetContentOffset: UnsafeMutablePointer<CGPoint>) {
+        let x = targetContentOffset.pointee.x
+        if x < currentPage.view.frame.minX, let left = leftPage {
+            rightPage?.unembed()
+            leftPage = nil
+            rightPage = currentPage
+            currentPage = left
+            notifyUpdated()
+        } else if x >= currentPage.view.frame.maxX, let right = rightPage {
+            leftPage?.unembed()
+            rightPage = nil
+            leftPage = currentPage
+            currentPage = right
+            scrollView.contentOffset.x -= scrollView.frame.width
+            targetContentOffset.pointee = CGPoint(x: x - scrollView.frame.width, y: 0)
```

Comment:
I think these two calls make sense to me, I'm just curious why you don't have to do them above when swiping right?

Reply:
You only need to mess with them when you remove the leftmost view (`leftPage?.unembed()`), which only happens in this branch. Dropping off the right view doesn't change where other views are positioned, so the scroll offset doesn't need to change either.

Reply:
Ok that makes sense.